tests/test_scan_services.py

```diff
@@ -0,0 +1,215 @@
+                logger.debug(f"simulate exception on request: {req.hex()} -> {arg!r}")
+                raise arg
+
+    async def mock_write(
+        self,
+        data: bytes,
+        timeout: float | None = None,
+        tags: list[str] | None = None,
+    ) -> int:
+        logger.debug(f"write out of request_unsafe: {data.hex()}")
+        self.req_pdu = data
+        return len(data)
+
+    async def mock_close(self) -> None:
+        logger.debug("mock_close")
+
+    async def mock_reconnect(self, timeout: float | None = None) -> AsyncMock:
+        logger.debug("mock_reconnect")
+        return await self.mock_connect(self.args.target)
+
+    async def mock_request_unsafe(
+        self, data: bytes, timeout: float | None = None, tags: list[str] | None = None
+    ) -> bytes:
+        res, duration = await self.transport.handle_request(data)
+        if not res:
+            raise TimeoutError
+        return await self.send_response(data, res, tags)
+
+    async def mock_connect(
+        self, target: str | TargetURI, timeout: float | None = None
+    ) -> AsyncMock:
+        con = AsyncMock()
+        con.close.side_effect = self.mock_close
+        con.write.side_effect = self.mock_write
+        con.read.side_effect = self.mock_read
+        con.reconnect.side_effect = self.mock_reconnect
+        con.request_unsafe.side_effect = self.mock_request_unsafe
+        return con
+
+    @pytest.mark.asyncio
+    @pytest.fixture
+    @patch("gallia.transports.unix.UnixTransport.connect")
+    @patch("gallia.command.uds.UDSScanner.configure_class_parser")
+    @patch("gallia.commands.scan.uds.services.ServicesScanner.configure_parser")
+    async def setup(
+        self,
+        mock_configure_parser: MagicMock,
+        mock_configure_class_parser: MagicMock,
+        mock_transport: MagicMock,
+    ) -> None:
+        await self.init()
+        mock_transport.side_effect = self.mock_connect
+        await self.scanner.setup(self.args)  # type: ignore
+
+    @pytest.mark.asyncio
+    @pytest.fixture
+    async def teardown(self) -> None:
+        await self.scanner.teardown(self.args)  # type: ignore
+
+
+class TestScanServices(ScanServicesMock):
+    @pytest.mark.asyncio
+    async def test_1(self, setup: None, teardown: None) -> None:
+        """Basic test of Service Scan against vECU.
+        Scanner should scan Session 1 and skip all SIDs expect 0x27.
+        """
+        assert self.scanner is not None
+        await self.scanner.main(self.args)  # type: ignore
+
+        assert len(self.scanner.found) == 1
+        assert len(self.scanner.found[1]) == 1
+        assert 0x27 in self.scanner.found[1]
+
+    @pytest.mark.asyncio
+    async def test_2(self, setup: None, teardown: None) -> None:
+        """Test TimeoutError on Service Scan.
+        On SID 0x27 a TimeoutError is simulated.
+        The Scanner should return this in the result.
+        """
+        assert self.scanner is not None
+
+        async def handle(
```

Comment:
As discussed via phone: This will be replaced via a table with TypedDicts, as suggested by russ cox: https://research.swtch.com/testing. Especially his tips 1, 5, 8, and 11 are relevant here.

Your draft from #495 could be used for a starting point.